mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp

```diff
@@ -398,8 +404,12 @@ struct TosaValidation : public tosa::impl::TosaValidationBase<TosaValidation> {
     }
   }
 
+  bool CheckVariable(Operation *op);
+  bool CheckVariableReadOrWrite(Operation *op);
+
   SmallVector<std::function<LogicalResult(Operation *)>> const_checkers;
   tosa_level_t tosa_level;
+  std::unordered_map<std::string, mlir::Type> variables_map;
```

Comment:
Use an Attribute as key: we'll save many copies and get more efficient hashing.

Also use a DenseMap unless you have some very specific iterator invalidation concerns, in which case you should document it.

Reply:
done

Reply:
I guess there was a misunderstanding here, now the code uses StringAttr*, but I think @joker-eph meant to use StringAttr. With --config=asan there are now failing tests from the newly added tests, I think we cannot rely on the address being the same if the string inside the StringAttr is the same.

Reply:
thanks for pointing out the --config=asan to reproduce the build failure.
working on a fix now.

Reply:
I already tried that just replacing StringAttr* in the DenseHashMap with StringAttr, and accessing the key without "&" dereference, makes the test pass with --config=asan. So these small changes should be enough :)

Reply:
I have landed this fix now in https://github.com/llvm/llvm-project/commit/7f2dd2da99371ef5b281834b604d251f3112cb23

Reply:
@Tai78641 The variable naming is also not aligned with our standard, it should be `variablesMap` here I believe. Maybe you could do a pass over this file (and similar)?

Reply:
will do.